common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1775,13 +1783,14 @@ public int selectApplicationProtocol(byte[] protocols) {
 
     @Override
     public String getApplicationProtocol() {
-        return SSLUtils.toProtocolString(ssl.getApplicationProtocol());
+        return provideAfterHandshakeSession().getApplicationProtocol();
     }
 
     @Override
     public String getHandshakeApplicationProtocol() {
         synchronized (ssl) {
-            return state == STATE_HANDSHAKE_STARTED ? getApplicationProtocol() : null;
+            return state >= STATE_HANDSHAKE_STARTED && state < STATE_READY
```

Comment:
Sorry, this change is unrelated but it's a bug I noticed while adding provideAfterHandshakeSession.

See https://developer.android.com/reference/javax/net/ssl/SSLSocket#getHandshakeSession()

Reply:
Actually my bad, this is *meant* to behave differently from gethandshakeSession()